### ISSUE TITLE
Output wall clock time per iteration?

### ISSUE THREAD

**Issue by snapdeb** [JuliaOpt/Ipopt.jl#178]
Hello,

I would like to know if there is an option to output the elapsed wall clock time per iteration? If there is no such function, i would like to modify the source code myself to add it. Can you point me to the file which i would have to edit?

**Comment by blegat** [comment]
It is done in https://github.com/JuliaOpt/Ipopt.jl/pull/170

**Comment by snapdeb** [comment]
Nice, also thanks for your fast answer :)

Do i have to wait until the commit is merged or can i possibly use the branch with this commit right now? I am also thinking of modifying the source code until this feature is added.

Currently, when i try to use the branch with the commit, I get:
`(v1.1) pkg> add https://github.com/JuliaOpt/Ipopt.jl.git#ml/moi09                                                                                              Updating git-repo `https://github.com/JuliaOpt/Ipopt.jl.git`                                                                                                Resolving package versions...                                                                                                                               ERROR: Unsatisfiable requirements detected for package MathOptInterface [b8f27783]:                                                                           MathOptInterface [b8f27783] log:                                                                                                                             ├─possible versions are: [0.1.0, 0.2.0, 0.3.0, 0.4.0-0.4.1, 0.5.0-0.5.1, 0.6.0-0.6.4, 0.7.0, 0.8.0-0.8.4] or uninstalled                                     └─restricted to versions 0.9.0-0.9 by Ipopt [b6b21f68] — no versions left                                                                                      └─Ipopt [b6b21f68] log:                                                                                                                                        ├─possible versions are: 0.6.0 or uninstalled                                                                                                                └─Ipopt [b6b21f68] is fixed to version 0.6.0   `

**Comment by blegat** [comment]
You need use the master branch of MathOptInterface and this branch of JuMP: https://github.com/JuliaOpt/JuMP.jl/pull/2003 if you want to use the branch of #170

**Comment by snapdeb** [comment]
Thank you. It is working, but not for my use case (with PowerModels.jl). I think this issue can be closed for now, if you don't want to leave it open for further discussions.
![grafik](https://user-images.githubusercontent.com/26022993/62453031-c2dd9600-b771-11e9-8ab0-baf1e6a0b22d.png)

**Comment by jchen975** [comment]
> Thank you. It is working, but not for my use case (with PowerModels.jl). I think this issue can be closed for now, if you don't want to leave it open for further discussions.
> ![grafik](https://user-images.githubusercontent.com/26022993/62453031-c2dd9600-b771-11e9-8ab0-baf1e6a0b22d.png)

Hi there, were you able to resolve this issue on PowerModels.jl? I'm doing a large number of `run_ac_pf()` in parallel and I would like to know **each run's number of iterations** and **each iteration's elapsed time**. It would be a pain to parse the output file because of the `From worker x` outputs and all the other information. Any advice is appreciated!

**Comment by lschmelting** [comment]
Hi @jchen975,

The following screenshot shows the current state. The Ipopt#master branch is needed since the wall clock time per iteration is included there. Also i chose the PowerModels#master branch since it has fever unsatisfiable requirements (since it is more up-to-date).
![grafik](https://user-images.githubusercontent.com/49401816/63272414-47fd9a80-c29c-11e9-8060-513adac16f25.png)
The problem is that Ipopt#master depends on MathOptInterface v0.9.0, while PowerModels depends on MathOptInterface v0.8.0. Thus we need to wait until PowerModels receives a new update, or assist in the development if at all possible. You could create an issue on PowerModels Github Repo page, but i doubt it would get a high priority since it is such a minor change.

**Comment by jchen975** [comment]
@lschmelting  Hey, thanks for your reply and opening the issue there. I suppose before that is updated, knowing the number of iterations in each `run_ac_pf` will suffice. Do you know if there is a function for that? I'm still pretty new to using JuliaOpt, but does `BarrierIterations()` have anything to do with what I want? 

Alternatively, I would have to parse the Ipopt output, but I really don't want to do that if there's something I can use directly.

**Comment by lschmelting** [comment]
When you call JuMP you should use the default `print_level=5`. Some timing_statistics are shown when using `print_timing_statistics="yes"` also you can specify an output_file. I am not sure which option is responsible for outputting the number of iterations. But it is incuded in the output file. Please try it out and report back:

`JuMP.with_optimizer(Ipopt.Optimizer, print_level=5, print_timing_statistics="yes", output_file="C:\\Users\\lschmelting\\workspace\\Ipopt_Output")`

**Comment by jchen975** [comment]
> When you call JuMP you should use the default `print_level=5`, which if I remember correctly, includes the number of iterations and some info for each iteration. Some additonal timing statistics are shown when using `print_timing_statistics="yes"`. Additionally, you can specify an output_file.
> 
> `JuMP.with_optimizer(Ipopt.Optimizer, print_level=5, print_timing_statistics="yes", output_file="C:\\Users\\lschmelting\\workspace\\Ipopt_Output")`

Yes, and actually setting `print_level=3` prints the number of iterations without further details, which I think is what I need for now. However, specifying an output file still requires further inspection or parsing to find the itr # information. Is there a better way to do that directly from PowerModels/JuMP/Ipopt?

**Comment by lschmelting** [comment]
I don't know of a better way, maybe someone else does.

**Comment by jchen975** [comment]
This short python script does the job. It's actually easier than I expected, but it would be nice to have a function to get #itr instead of having to parse the output file. 
```
filepath = "./test.log"  # change this to the directory containing the output.log
key = "Number of Iterations...."

itr = []
with open(filepath) as fp:
	s = fp.readline()
	while s:
		l = [ss.strip() for ss in s.split(":")]
		if len(l) >= 2 and l[-2] == key:
			itr.append(int(l[-1]))
		s = fp.readline()
```

**Comment by mlubin** [comment]
Unless you can find somewhere to extract this information from [Ipopt's API](https://github.com/coin-or/Ipopt/blob/master/Ipopt/src/Interfaces/IpStdCInterface.h), there's not really much we can do on the Julia side.

**Comment by lschmelting** [comment]
@jchen975: PowerModels is now usable with the Ipopt master branch: `add Ipopt#master PowerModels#master`.

I don't see a wall clock time per iteration output, though.
I use Ipopt like this: `ipopt_solver = JuMP.with_optimizer(Ipopt.Optimizer, print_level=5, print_timing_statistics="yes", output_file="C:\\Users\\lschmelting\\workspace\\test")`.

Any hints?

**Comment by dpo** [comment]
In the mean time, you could pass a callback and measure time yourself.
